src/main/scala/com/microsoft/hyperspace/index/Index.scala

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright (2021) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index
+
+import com.fasterxml.jackson.annotation.{JsonSubTypes, JsonTypeInfo}
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+/**
+ * IndexConfig specifies the configuration of an index.
+ *
+ * Index implementations should extend this trait and make the class visible
+ * to users. Users can use the class when creating an index with the
+ * [[com.microsoft.hyperspace.Hyperspace.createIndex()]] API.
+ */
+trait IndexConfig {
+
+  /**
+   * Returns the name of the index.
+   *
+   * Index names are case-insensitive. An index name should not contain
+   * disallowed characters for the file system, as index data is stored
+   * in the file system with the index name in the path.
+   */
+  def indexName: String
+
+  /**
+   * Returns the column names which the index references.
+   */
+  def referencedColumns: Seq[String]
+
+  /**
+   * Creates [[IndexMetadata]] from this configuration.
+   *
+   * @param spark Spark session
+   * @param columnResolver Used to resolve column names against the source plan
+   * @return [[IndexMetadata]] that can be used to create an index
+   */
+  def createMetadata(spark: SparkSession, columnResolver: ColumnResolver): IndexMetadata
+}
+
+/**
+ * Represents parameters that define an index.
+ *
+ * While [[IndexConfig]] is designed for users to specify parameters for an
+ * index to create, [[IndexMetadata]] includes more information filled with
+ * values determined by the framework and the implementation that can be used
+ * directly to create an index.
+ */
+trait IndexMetadata {
```

Comment:
Please extract this trait in a separate file.

Reply:
Done